[REVIEW] Move java support to libcudf++ APIs

revans2 opened the pull request:
This changes the backend of the java API to the new cudf++ API.

This work was done in part by a lot of different people over a lengthy time.  It also includes a new new java APIs for functionality available in the libcudf++ APIs that were not available in the original API.  As much as possible the java APIs have remained unchanged.

razajafri commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 ## Improvements
 
+- PR #3909 Mova java backend to libcudf++
```
sb Move

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -19,75 +19,119 @@
 
 #include <unordered_set>
```
This might not be needed anymore, since the code that used `unordered_set` has been deleted.

revans2 commented:
@razajafri I fixed the type
@jlowe I went through all of the includes for the JNI code and removed as much as I could.  There were several headers that need not be included or were redundant.

jlowe commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -16,10 +16,7 @@
 #include <cudf/binaryop.hpp>
-#include <cudf/column/column.hpp>
 #include <cudf/column/column_factories.hpp>
-#include <cudf/column/column_view.hpp>
```
I'm hesitant to remove headers that are directly providing a type or API that we are using in our code.  Clearly we need this particular header and some others here that were removed.  We must be getting them via some other headers we're including, or it wouldn't compile.  However in the future if those other headers stop including these headers, because they start forward declaring the types move their implementation out of headers, then our builds start breaking.

I'm all for removing headers that are not referenced in the code, but I think we should continue to include headers that we are using directly even if it would compile without them due to other header dependencies.  Not must-fix, just my personal preference.  Curious what others think.

revans2 replied:
Yes I realize that.  The cleanup of the APIs had thought go into it.  I didn't just remove everything that could be removed and called it good.  In this case the column_view, and in most of the other cases with table, table_view, column, and strings_column_view.  They are things that the APIs we use use directly.  How can I sort a table without a table_view?

Yes in the future someone may find a way to make a table_view a pure forward declaration, but with how it is used I doubt it will ever happen.
